chemprop/utils.py

```diff
@@ -190,80 +202,171 @@ def load_frzn_model(model: torch.nn,
-            model_encoder_param_names = [['encoder.encoder.'+str(mol_num)+'.W_i.weight', 'encoder.encoder.'+str(mol_num)+'.W_h.weight', 'encoder.encoder.'+str(mol_num)+'.W_o.weight', 'encoder.encoder.'+str(mol_num)+'.W_o.bias'] for mol_num in range(current_args.number_of_molecules)]
-            model_encoder_param_names = [item for sublist in model_encoder_param_names for item in sublist]
-            for loaded_param_name,model_param_name in zip(loaded_encoder_param_names,model_encoder_param_names):
-                model_state_dict = overwrite_state_dict(loaded_param_name,model_param_name,loaded_state_dict,model_state_dict)
-        
-        if current_args.frzn_ffn_layers > 0: # Duplicate encoder from frozen checkpoint and overwrite all encoders + FFN layers
-            raise Exception ('Number of molecules in checkpoint_frzn must be equal to current model for ffn layers to be frozen')
-            
-    elif (loaded_args.number_of_molecules>1 )& (current_args.number_of_molecules>1):
-        if (loaded_args.number_of_molecules) !=( current_args.number_of_molecules):
-            raise Exception('Number of molecules in checkpoint_frzn ({}) must match current model ({}) OR equal to 1.'.format(loaded_args.number_of_molecules,current_args.number_of_molecules))
-        
+                model_state_dict = overwrite_state_dict(
+                    param_name, param_name, loaded_state_dict, model_state_dict
+                )
+        if (
+            current_args.checkpoint_frzn is not None
+            and not current_args.freeze_first_only
+            and current_args.frzn_ffn_layers <= 0
+        ):  # Duplicate encoder from frozen checkpoint and overwrite all encoders
+            loaded_encoder_param_names = [
+                "encoder.encoder.0.W_i.weight",
+                "encoder.encoder.0.W_h.weight",
+                "encoder.encoder.0.W_o.weight",
+                "encoder.encoder.0.W_o.bias",
+            ] * current_args.number_of_molecules
+            model_encoder_param_names = [
+                [
+                    (
+                        f"encoder.encoder.{mol_num}.W_i.weight",
+                        f"encoder.encoder.{mol_num}.W_h.weight",
+                        f"encoder.encoder.{mol_num}.W_o.weight",
+                        f"encoder.encoder.{mol_num}.W_o.bias",
+                    )
+                ]
+                for mol_num in range(current_args.number_of_molecules)
+            ]
+            model_encoder_param_names = [
+                item for sublist in model_encoder_param_names for item in sublist
+            ]
+            for loaded_param_name, model_param_name in zip(
+                loaded_encoder_param_names, model_encoder_param_names
+            ):
+                model_state_dict = overwrite_state_dict(
+                    loaded_param_name, model_param_name, loaded_state_dict, model_state_dict
+                )
```

Comment:
This block is difficult to read with the new formatting. Not that the model_encoder_param_names was easy to read before either.

Reply:
The pattern `[item for sublist in model_encoder_param_names for item in sublist]` shows up several times in the frozen checkpoint sections and seems like something that could be combined with what it's unpacking in a more clear way perhaps.

Reply:
@cbilodeau2 can you help us understand this code block?